File: #48_Rotate_Image/Solution.py
class Solution:
    # @param {integer[][]} matrix
    # @return {void} Do not return anything, modify matrix in-place instead.
    def rotate(self, matrix):
        if not matrix or len(matrix) == 1:
        	return
        n = len(matrix)
        for x in range(n / 2):
        	for y in range(x, n - 1 - x):
        		i, j = x, y
        		a = matrix[i][j]
        		b = matrix[j][n - 1 - i]
        		c = matrix[n - 1 - i][n - 1 - j]
        		d = matrix[n - 1 - j][i]
        		matrix[i][j] = d
        		matrix[j][n - 1 - i] = a
        		matrix[n - 1 - i][n - 1 - j] = b
        		matrix[n - 1 - j][i] = c

sol = Solution()
t = [[0, 1 ,2],
	 [3, 4, 5],
	 [6, 7, 8]]
sol.rotate(t)
print t

File: #48_Rotate_Image/Solution.java
public class Solution {
    public void rotate(int[][] matrix) {
        if (matrix == null || matrix.length == 0 || matrix[0].length != matrix.length) return;
        int n = matrix.length;
        for (int i = 0; i < n / 2; i++) {
        	for (int j = i; j < n - i - 1; j++) {
        		// (i, j) -> (j, n - i - 1)
        		int tmp = matrix[j][n - i - 1];
        		matrix[j][n - i - 1] = matrix[i][j];
        		// (j, n - i - 1) -> (n - i - 1, n - j - 1)
        		int tmp2 = matrix[n - i - 1][n - j - 1];
        		matrix[n - i - 1][n - j - 1] = tmp;
        		tmp = tmp2;
        		// (n - i - 1, n - j - 1) -> (n - j - 1, i)
        		tmp2 = matrix[n - j - 1][i];
        		matrix[n - j - 1][i] = tmp;
        		tmp = tmp2;
        		// (n - j - 1, i) -> (i, j)
        		matrix[i][j] = tmp;
        	}
        }
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	int[][] matrix = {
    		{1, 2, 3, 4},
    		{5, 6, 7, 8},
    		{9, 10, 11, 12},
    		{13, 14, 15, 16}
    	};
    	sol.rotate(matrix);
    	for (int i = 0; i < matrix.length; i++) {
    		for (int j = 0; j < matrix.length; j++) {
    			System.out.print(matrix[i][j] + " ");
    		}
    		System.out.println();
    	}
    }
}